test/e2e/storage/sanity.go

```diff
@@ -166,3 +267,176 @@ func getDaemonSet(cs clientset.Interface, setName string) *appsv1.DaemonSet {
+		&csi.NodeStageVolumeRequest{
+			VolumeId: vol.GetVolumeId(),
+			VolumeCapability: &csi.VolumeCapability{
+				AccessType: &csi.VolumeCapability_Mount{
+					Mount: &csi.VolumeCapability_MountVolume{},
+				},
+				AccessMode: &csi.VolumeCapability_AccessMode{
+					Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+				},
+			},
+			StagingTargetPath: sc.StagingPath,
+			VolumeContext:     vol.GetVolumeContext(),
+			PublishContext:    conpubvol.GetPublishContext(),
+		},
+	)
+	framework.ExpectNoError(err, "node stage volume")
+	Expect(nodestagevol).NotTo(BeNil())
+
+	// NodePublishVolume
+	By("publishing the volume on a node")
+	nodepubvol, err := c.NodePublishVolume(
+		context.Background(),
+		&csi.NodePublishVolumeRequest{
+			VolumeId:          vol.GetVolumeId(),
+			TargetPath:        sc.TargetPath + "/target",
+			StagingTargetPath: sc.StagingPath,
+			VolumeCapability: &csi.VolumeCapability{
+				AccessType: &csi.VolumeCapability_Mount{
+					Mount: &csi.VolumeCapability_MountVolume{},
+				},
+				AccessMode: &csi.VolumeCapability_AccessMode{
+					Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+				},
+			},
+			VolumeContext:  vol.GetVolumeContext(),
+			PublishContext: conpubvol.GetPublishContext(),
+		},
+	)
+	framework.ExpectNoError(err, "node publish volume")
+	Expect(nodepubvol).NotTo(BeNil())
+
+	return nid.GetNodeId()
+}
+
+func unpublishVolume(s csi.ControllerClient, c csi.NodeClient, sc *sanity.SanityContext, vol *csi.Volume, nodeID string) {
+	var err error
+
+	By("cleaning up calling nodeunpublish")
+	nodeunpubvol, err := c.NodeUnpublishVolume(
+		context.Background(),
+		&csi.NodeUnpublishVolumeRequest{
+			VolumeId:   vol.GetVolumeId(),
+			TargetPath: sc.TargetPath + "/target",
+		})
+	framework.ExpectNoError(err, "node unpublish volume")
+	Expect(nodeunpubvol).NotTo(BeNil())
+
+	By("cleaning up calling nodeunstage")
+	nodeunstagevol, err := c.NodeUnstageVolume(
+		context.Background(),
+		&csi.NodeUnstageVolumeRequest{
+			VolumeId:          vol.GetVolumeId(),
+			StagingTargetPath: sc.StagingPath,
+		},
+	)
+	framework.ExpectNoError(err, "node unstage volume")
+	Expect(nodeunstagevol).NotTo(BeNil())
+}
+
+func deleteVolume(s csi.ControllerClient, vol *csi.Volume) {
+	var err error
+
+	By("cleaning up deleting the volume")
+	_, err = s.DeleteVolume(
+		context.Background(),
+		&csi.DeleteVolumeRequest{
+			VolumeId: vol.GetVolumeId(),
+		},
+	)
+	framework.ExpectNoError(err, "delete volume %s", vol.GetVolumeId())
+}
+
+var unreachable = v1.Taint{Key: "node.kubernetes.io/unreachable", Effect: "NoSchedule"}
+
+// restartNode works only for one of the nodes in the QEMU virtual cluster.
+// It does a hard poweroff via SysRq and relies on Docker to restart the
+// "failed" node.
+func restartNode(cs clientset.Interface, nodeID string) {
+	if !regexp.MustCompile(`worker\d+$`).MatchString(nodeID) {
+		framework.Skipf("node %q not one of the expected QEMU nodes (worker<number>))", nodeID)
+	}
```

Comment:
Is this function striking the right balance between "don't break a real system" and "ensure that it runs"?

It's using the node ID to determine the node number because it needs to find the ssh script for it. Should we ever changing the naming convention, then this will break and the test will get skipped.